---
title: "The sequencing_set R6 class"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Preamble

The `sequencing_set` is an object that describes a collection of sequencing
information pulled from a folder of basecalled sequence data. In the first
vignette in the documentation series, we reviewed how sequence files could
be identified and indexed using the `basecalled_folder` class. This vignette
focuses on the extraction of value from the sequence information.

## Getting started with the canned dataset

The `rtqc` package is distributed with some simplified data. In the first 
vignette we prepared a sequence_set with a canonical logic that included

```{r load_rtqc}
library(rtqc)

dorado_fastq_path <- get_bundled_path("fastq_pass")
bf <- basecalled_folder$new(dorado_fastq_path)
bf$index(threads=3)
my_sequence_set <- bf$as_sequence_set()
print(my_sequence_set)
```

In the block above, the output from `my_sequence_set` shows that the data
is currently being indexed. The speed of indexing depends on the size of the
sequence files (how many sequence reads, and length of individual sequences).
There is a risk that if we try to do something with these sequences straight
away that there will be no indexed data. The indexing runs as a background
process - we can use the `wait_until_data` method to pause this code until
there is some actionable data ...

```{r wait_until_data}
my_sequence_set$wait_until_data() # blocking step ... 
print(my_sequence_set)
my_sequence_set$get_parquet_pairs()
```

The final facet of this object is that we can extract the aggregate data. This
will load the sequence characteristics from the bundled `.parquet` files - the
files are not loaded into a monolith in the current version of the code; if you
look into the project's older branches there was some dubious rust code to
perform file aggregation.

```{r getdata}
my_sequence_set$get_data()
```

Of course, if we have gotten this far with an analysis, then there is a chance
that some additional files will have been indexed. This can be checked with
the `sync()` method that will return `TRUE` only if there are new sequence
elements that might be considered.

```{r sync}
my_sequence_set$sync()
```

A collection of sequence data characteristics alone is a rather nebulous 
capability. It is more interesting to present these characteristics in a tabular
in graphical form. This is better achieved using the `sequence_set_summary`
class. This can be prepared with

```{r sequence_set_summary}
summary <- my_sequence_set$as_summary()
print(summary)
```
